scripts/llext_link_helper.py

```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
+# SPDX-License-Identifier: BSD-3-Clause
+
+# We need to calculate ELF section addresses of an LLEXT module and use them to
+# run the linker. We could just use Python to calculate addresses and pass them
+# back to cmake to have it call the linker. However, there doesn't seem to be a
+# portable way to do that. Therefore we pass the linker path and all the command
```

Comment:
@tejlmand , any idea?

I think we want this to be done in Zephyr for any `llext` eventually, not just SOF. @lyakh this problem is not SOF specific or is it? Either way state it in this comment so we know where the solution for this should live eventually.

Reply:
@marc-hb it is SOF-specific. It isn't needed for generic llext loading on Xtensa under Zephyr. It's only needed to match SOF's loadable module / manifest design

Reply:
Why can't we do it in linker scripts?

Reply:
> Why can't we do it in linker scripts?

@pjdobrowolski why should we? If you use a linker script you need to define everything, a complete module linking layout. I see no need for that. The only thing that we need is a fixed memory location where we want to map respective modules, that's the only parameter I'm using. The rest is calculated automatically. In fact ideally even the load address shouldn't be needed. Think about OS executables, libraries, Linux kernel modules, etc. The OS automatically finds a location to copy the code and the data to and automatically selects addresses to map it to.

Reply:
Because it keeps all memory locations and operations in one place. I understand that it is easier to do it but FW is a bit more complex that software and adding python to it doesn't help.

Reply:
> FW is a bit more complex that software  and adding python to it doesn't help.

This is quite vague and cryptic, can you elaborate? Python has great support for ELF. It is portable, popular, interactive and high-level. So it's hard to find another tool that checks so many boxes for such a task. It's also consistent with Zephyr which uses all over the place for similar jobs.

Maybe your comment wasn't about the choice of the Python language specifically but more generally about computing addresses versus hardcoding them?

> Because it keeps all memory locations and operations in one place

This is a bit confusing too: addresses dynamically computed are not kept in a different place... would it help to make sure they get printed and saved?

Reply:
BTW there is on-going work to add a post-processing step in CMake:
- https://github.com/zephyrproject-rtos/zephyr/pull/67997

Reply:
> FW is a bit more complex that software...

Depends how you look at it
https://en.wikipedia.org/wiki/Address_space_layout_randomization